server/src/main/java/com/cloud/api/query/QueryManagerImpl.java

```diff
@@ -3423,12 +3432,22 @@ else if (!template.isPublicTemplate() && caller.getType() != Account.ACCOUNT_TYP
             return uniqueTmplPair;
         }
         List<TemplateJoinVO> uniqueTmpls = uniqueTmplPair.first();
-        String[] tzIds = new String[uniqueTmpls.size()];
         int i = 0;
-        for (TemplateJoinVO v : uniqueTmpls) {
-            tzIds[i++] = v.getTempZonePair();
+        List<TemplateJoinVO> vrs = null;
+        if (showUnique) {
+            Long[] tzIds = new Long[uniqueTmpls.size()];
+            for (TemplateJoinVO v : uniqueTmpls) {
+                tzIds[i++] = v.getId();
+            }
+            vrs = _templateJoinDao.findByDistinctIds(tzIds);
+        } else {
+            String[] tzIds = new String[uniqueTmpls.size()];
+            for (TemplateJoinVO v : uniqueTmpls) {
+                tzIds[i++] = v.getTempZonePair();
+            }
+            vrs = _templateJoinDao.searchByTemplateZonePair(showRemovedTmpl, tzIds);
```

Comment:
Unit testing for such a huge/complex method is not easy. However, it would be great to address small pieces.

With that said, lines 3436 - 3448 could be extracted into a method that would allow testing this specific piece of code.